src/app/cli/src/services/odf_server/login_service.rs

```diff
@@ -0,0 +1,282 @@
+        fields(
+            body = %body,
+        )
+    )]
+    async fn post_handler(
+        axum::extract::State(response_tx): axum::extract::State<
+            tokio::sync::mpsc::Sender<FrontendLoginCallbackResponse>,
+        >,
+        body: String,
+    ) -> impl axum::response::IntoResponse {
+        let response_result = serde_json::from_str::<FrontendLoginCallbackResponse>(body.as_str());
+        match response_result {
+            Ok(response) => {
+                response_tx.send(response).await.unwrap();
+
+                axum::response::Response::builder()
+                    .status(200)
+                    .body("{}".to_string())
+                    .unwrap()
+            }
+            Err(e) => axum::response::Response::builder()
+                .status(400)
+                .body(e.to_string())
+                .unwrap(),
+        }
+    }
+
+    fn initialize_cli_web_server(
+        &self,
+        server_frontend_url: &Url,
+        response_tx: tokio::sync::mpsc::Sender<FrontendLoginCallbackResponse>,
+    ) -> WebServer {
+        let addr = SocketAddr::from((IpAddr::V4(Ipv4Addr::new(127, 0, 0, 1)), 0));
+
+        let bound_addr = hyper::server::conn::AddrIncoming::bind(&addr).unwrap_or_else(|e| {
+            panic!("error binding to {}: {}", addr, e);
+        });
+
+        let redirect_url = format!(
+            "{}?callbackUrl=http://{}/",
+            server_frontend_url.join("/v/login").unwrap(),
+            bound_addr.local_addr().to_string()
+        );
+
+        let app = axum::Router::new()
+            .route(
+                "/",
+                axum::routing::get(|| async move {
+                    axum::response::Redirect::permanent(redirect_url.as_str())
+                })
+                .post(Self::post_handler),
+            )
+            .layer(
+                tower::ServiceBuilder::new()
+                    .layer(tower_http::trace::TraceLayer::new_for_http())
+                    .layer(
+                        tower_http::cors::CorsLayer::new()
+                            .allow_origin(tower_http::cors::Any)
+                            .allow_methods(vec![http::Method::GET, http::Method::POST])
+                            .allow_headers(tower_http::cors::Any),
+                    ),
+            )
+            .with_state(response_tx);
+
+        axum::Server::builder(bound_addr).serve(app.into_make_service())
+    }
+
+    fn open_web_browser(&self, cli_web_server_url: &String) {
+        tracing::info!("HTTP server is listening on: {}", cli_web_server_url);
+
+        if self.output_config.is_tty
+            && self.output_config.verbosity_level == 0
+            && !self.output_config.quiet
+        {
+            eprintln!(
+                "{}\n  {}",
+                s("HTTP server is listening on:").green().bold(),
+                s(&cli_web_server_url).bold(),
+            );
+            eprintln!("{}", s("Use Ctrl+C to stop the server").yellow());
```

Comment:
1) Can we keep all output in the command instead of the service?

2) Do we want to show user anything about the HTTP server running and how to stop it? I would expect a message like this instead:
```
Please open this URL in the browser to login https://....
This process will exit automatically after receiving the login result.
```

Reply:
1. Moved reporting into command via a listener callback.
2. Improved as recommended:
![image](https://github.com/kamu-data/kamu-cli/assets/8875909/3ccd939b-c756-4a2d-951e-8d5499caa9c4)